src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionContext.cs

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Net.Quic;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Connections.Abstractions.Features;
+using Microsoft.AspNetCore.Connections.Features;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Internal
+{
+    internal class QuicConnectionContext : TransportConnection, IQuicStreamListenerFeature, IQuicCreateStreamFeature
+    {
+        private QuicConnection _connection;
+        private readonly QuicTransportContext _context;
+        private readonly IQuicTrace _log;
+
+        private ValueTask _closeTask;
+
+        public QuicConnectionContext(QuicConnection connection, QuicTransportContext context)
+        {
+            _log = context.Log;
+            _context = context;
+            _connection = connection;
+            Features.Set<ITlsConnectionFeature>(new FakeTlsConnectionFeature());
+            Features.Set<IQuicStreamListenerFeature>(this);
+            Features.Set<IQuicCreateStreamFeature>(this);
+
+            _log.NewConnection(ConnectionId);
+        }
+
+        public ValueTask<ConnectionContext> StartUnidirectionalStreamAsync()
+        {
+            var stream = _connection.OpenUnidirectionalStream();
+            return new ValueTask<ConnectionContext>(new QuicStreamContext(stream, this, _context));
+        }
+
+        public ValueTask<ConnectionContext> StartBidirectionalStreamAsync()
+        {
+            var stream = _connection.OpenBidirectionalStream();
+            return new ValueTask<ConnectionContext>(new QuicStreamContext(stream, this, _context));
+        }
+
+        public override async ValueTask DisposeAsync()
+        {
+            try
+            {
+                if (_closeTask != default)
+                {
+                    await _connection.CloseAsync(errorCode: 0);
```

Comment:
Nit: It's probably worth storing this in `_closeTask` to protect against multiple calls to DisposeAsync.